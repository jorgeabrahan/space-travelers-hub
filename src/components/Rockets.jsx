import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../redux/rockets/rocketsSlice';
import Rocket from './Rocket';

function Rockets() {
  const dispatch = useDispatch();
  const { rockets, status } = useSelector((store) => store.rockets);
  useEffect(() => {
    /* Validation to fetch data just when the page loads */
    if (rockets.length > 0) return;
    if (status === 'failed') return;
    dispatch(fetchRockets());
  }, [dispatch]);

  const rocketsList = rockets.map(({
    id, name, type, images, description,
  }) => (
    <Rocket
      key={id}
      name={name}
      type={type}
      image={images[0]}
      description={description}
    />
  ));

  return (
    <section className="center">
      {rocketsList}
    </section>
  );
}

export default Rockets;
