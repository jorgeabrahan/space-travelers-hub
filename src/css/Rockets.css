.rockets {
  display: grid;
  gap: 1.5rem;
}

.rocket {
  display: flex;
  gap: 1rem;
}

.rocket__image {
  width: 260px;
}

.rocket__title {
  font-size: 1.4rem;
  margin-bottom: 0.3rem;
}

.rocket__badge {
  background-color: #18a2b8;
  color: white;
  font-size: 0.85rem;
  font-weight: bold;
  padding: 0.1rem 0.3rem;
  border-radius: 5px;
  margin-right: 0.6rem;
}

.rocket__description {
  font-size: 1.05rem;
  line-height: 1.5rem;
  font-weight: 400;
  margin-bottom: 0.8rem;
}

.rocket__button {
  border: none;
  border-radius: 5px;
  font-size: 1.1rem;
  padding: 0.5rem 1rem;
}

.rocket__button.reserve {
  background-color: #007bff;
  color: white;
}

.rocket__button.reserved {
  background: none;
  border: 1px solid #606972;
  color: #606972;
}
