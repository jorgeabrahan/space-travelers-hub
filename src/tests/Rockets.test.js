/* eslint-disable import/no-extraneous-dependencies */
import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Rockets from '../components/Rockets';

describe('Tests for <Rockets /> component', () => {
  const title = 'Falcon 1';
  const titleNoSpace = title.replace(/\s/g, '');
  const description = 'Falcon description';
  const initialState = {
    rockets: [{
      id: '1',
      name: title,
      images: [`${titleNoSpace}.jpg`],
      description,
      reserved: false,
    }, {
      id: '2',
      name: title,
      images: [`${titleNoSpace}.jpg`],
      description,
      reserved: true,
    }],
    status: 'succeeded',
    error: '',
  };

  const rocketsSlice = createSlice({
    name: 'rockets',
    initialState,
  });

  const store = configureStore({
    reducer: {
      rockets: rocketsSlice.reducer,
    },
  });

  const wrapper = (component) => (
    <Provider store={store}>
      {component}
    </Provider>
  );

  it('Should render two rockets', () => {
    render(wrapper(<Rockets />));
    expect(screen.queryByTestId('rockets-container').children.length).toBe(2);
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(<Rockets />)).toJSON()).toMatchSnapshot();
  });
});
