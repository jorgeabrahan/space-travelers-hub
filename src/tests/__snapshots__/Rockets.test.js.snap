// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Rockets /> component Should match the snapshot 1`] = `
<section
  className="center rockets"
  data-testid="rockets-container"
>
  <div
    className="rocket"
  >
    <img
      alt="Falcon 1"
      className="rocket__image"
      loading="lazy"
      src="Falcon1.jpg"
    />
    <div
      className="rocket__content"
    >
      <h2
        className="rocket__title"
      >
        Falcon 1
      </h2>
      <p
        className="rocket__description"
      >
        Falcon description
      </p>
      <button
        className="rocket__button reserve"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="rocket"
  >
    <img
      alt="Falcon 1"
      className="rocket__image"
      loading="lazy"
      src="Falcon1.jpg"
    />
    <div
      className="rocket__content"
    >
      <h2
        className="rocket__title"
      >
        Falcon 1
      </h2>
      <p
        className="rocket__description"
      >
        <span
          className="rocket__badge"
        >
          Reserved
        </span>
        Falcon description
      </p>
      <button
        className="rocket__button reserved"
        onClick={[Function]}
        type="button"
      >
        Cancel Reservation
      </button>
    </div>
  </div>
</section>
`;
